File: api-certvalidation/csharp/src/find/FinderOCSPResponseSource.cs
using System.Collections.Generic;
using tr.gov.tubitak.uekae.esya.api.asn.ocsp;
using tr.gov.tubitak.uekae.esya.api.asn.x509;
using tr.gov.tubitak.uekae.esya.api.certificate.validation.find.ocsp;
using tr.gov.tubitak.uekae.esya.api.common.util;

namespace tr.gov.tubitak.uekae.esya.api.certificate.validation.find
{
    /**
     * @author ayetgin
     */
    public class FinderOCSPResponseSource<T> : ItemSource<EOCSPResponse> where T : Finder
    {
        private readonly ECertificate mSubject;
        private readonly ECertificate mIssuer;
        private readonly List<T> mFinders;
        private ItemSource<EOCSPResponse> mCurrentSource;
        private int mCurrentIndex;

        public FinderOCSPResponseSource(ECertificate aSubject, ECertificate aIssuer, List<T> aFinders)
        {
            mSubject = aSubject;
            mIssuer = aIssuer;
            mFinders = aFinders;
            mCurrentIndex = -1;
            mCurrentSource = null;
        }

        public bool open()
        {
            return reset();
        }

        public bool close()
        {
            return true;
        }

        public bool reset()
        {
            mCurrentIndex = -1;
            return true;
        }

        public EOCSPResponse nextItem()
        {
            EOCSPResponse pItem = null;
            if ((mCurrentSource != null) && ((pItem = mCurrentSource.nextItem()) != null))
                return pItem;

            while (true)
            {
                if ((++mCurrentIndex) >= mFinders.Count) return null;

                mCurrentSource = null;

                //mCurrentSource = ((OCSPResponseFinder)mFinders[mCurrentIndex]).findOCSPSource(mSubject, mIssuer);

                mCurrentSource = (mFinders[mCurrentIndex] as OCSPResponseFinder).findOCSPSource(mSubject, mIssuer);

                if ((mCurrentSource != null) && ((pItem = mCurrentSource.nextItem()) != null))
                    return pItem;
            }
        }

        public bool atEnd()
        {
            return (mCurrentIndex >= mFinders.Count);
        }
    }
}

File: api-certvalidation/src/tr/gov/tubitak/uekae/esya/api/certificate/validation/find/FinderOCSPResponseSource.java
package tr.gov.tubitak.uekae.esya.api.certificate.validation.find;

import tr.gov.tubitak.uekae.esya.api.asn.ocsp.EOCSPResponse;
import tr.gov.tubitak.uekae.esya.api.asn.x509.ECertificate;
import tr.gov.tubitak.uekae.esya.api.certificate.validation.find.ocsp.OCSPResponseFinder;
import tr.gov.tubitak.uekae.esya.api.common.ESYAException;
import tr.gov.tubitak.uekae.esya.api.common.util.ItemSource;

import java.util.List;

/**
 * @author ayetgin
 */
public class FinderOCSPResponseSource implements ItemSource<EOCSPResponse>
{
    private ECertificate mSubject, mIssuer;
    private List<? extends Finder> mFinders;
    ItemSource<EOCSPResponse> mCurrentSource;
    int mCurrentIndex;

    public FinderOCSPResponseSource(ECertificate aSubject, ECertificate aIssuer, List<? extends Finder> aFinders)
    {
        mSubject= aSubject;
        mIssuer = aIssuer;
        mFinders = aFinders;
        mCurrentIndex=-1;
        mCurrentSource=null;
    }

    public boolean open()
    {
        return reset();
    }

    public boolean close()
    {
        return true;
    }

    public boolean reset()
    {
        mCurrentIndex = -1;
        return true;
    }

    public EOCSPResponse nextItem() throws ESYAException
    {
        EOCSPResponse pItem = null;
        if ((mCurrentSource!=null) && ((pItem = mCurrentSource.nextItem())!=null))
            return pItem;

        while(true)
        {
            if ((++mCurrentIndex)>=mFinders.size()) return null;

            mCurrentSource = null;

            mCurrentSource = ((OCSPResponseFinder)mFinders.get(mCurrentIndex)).findOCSPSource(mSubject, mIssuer);

            if ((mCurrentSource!=null) && ((pItem = mCurrentSource.nextItem())!=null))
                return pItem;
        }

    }

    public boolean atEnd()
    {
        return (mCurrentIndex>=mFinders.size());
    }

}
